packages/opentelemetry-instrumentation-grpc/src/grpc.ts

```diff
@@ -0,0 +1,600 @@
+ * limitations under the License.
+ */
+
+import {
+  StatusCode,
+  context,
+  propagation,
+  Span,
+  SpanKind,
+  SpanOptions,
+  Status,
+} from '@opentelemetry/api';
+import { RpcAttribute } from '@opentelemetry/semantic-conventions';
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import type * as events from 'events';
+import type * as grpcTypes from 'grpc';
+import {
+  GrpcClientFunc,
+  GrpcInternalClientTypes,
+  GrpcInstrumentationConfig,
+  SendUnaryDataCallback,
+  ServerCallWithMeta,
+} from './types';
+import {
+  findIndex,
+  _grpcStatusCodeToOpenTelemetryStatusCode,
+  _grpcStatusCodeToSpanStatus,
+  _methodIsIgnored,
+} from './utils';
+import { VERSION } from './version';
+
+/** The metadata key under which span context is stored as a binary value. */
+export const GRPC_TRACE_KEY = 'grpc-trace-bin';
+
+/**
+ * Holding reference to grpc module here to access constant of grpc modules
+ * instead of just requiring it avoid directly depending on grpc itself.
+ */
+let grpcClient: typeof grpcTypes;
+
+export class GrpcInstrumentation extends InstrumentationBase<typeof grpcTypes> {
+  constructor(
+    protected _config: GrpcInstrumentationConfig & InstrumentationConfig = {}
+  ) {
+    super('@opentelemetry/instrumentation-grpc', VERSION, _config);
+  }
+
+  public setConfig(
+    config: GrpcInstrumentationConfig & InstrumentationConfig = {}
+  ) {
+    this._config = Object.assign({}, config);
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<typeof grpcTypes>(
+        'grpc',
+        ['1.*'],
+        (moduleExports, version) => {
+          this._logger.debug(`Applying patch for grpc@${version}`);
+          grpcClient = moduleExports;
+          if (isWrapped(moduleExports.Server.prototype.register)) {
+            this._unwrap(moduleExports.Server.prototype, 'register');
+          }
+          this._wrap(
+            moduleExports.Server.prototype,
+            'register',
+            this._patchServer() as any
+          );
+          // Wrap the externally exported client constructor
+          if (isWrapped(moduleExports.makeGenericClientConstructor)) {
+            this._unwrap(moduleExports, 'makeGenericClientConstructor');
+          }
+          this._wrap(
+            moduleExports,
+            'makeGenericClientConstructor',
+            this._patchClient()
+          );
+          return moduleExports;
+        },
+        (moduleExports, version) => {
+          if (moduleExports === undefined) return;
+          this._logger.debug(`Removing patch for grpc@${version}`);
+
+          this._unwrap(moduleExports.Server.prototype, 'register');
+        },
+        this._getInternalPatchs()
+      ),
+    ];
+  }
+
+  private _getInternalPatchs() {
+    return [
+      new InstrumentationNodeModuleFile<GrpcInternalClientTypes>(
```

Comment:
Nit: split this into 2 new internal function for better readability for example
```
private _getInternalPatchs() {
    return [
      this.patchClientVerTill1_6,
      this.patchClientVerAbove1_6,
    ]
```

Reply:
Since both patches were using the same duplicated code, i prefered to declare both `onPatch` and `onUnPatch` and just re-use them. PTAL